caffe2/proto/torch.proto

```diff
@@ -0,0 +1,599 @@
+
+  // Additional annotations.
+  repeated AttributeProto annotations = 17;
+
+  // Device info
+  optional DeviceOption device_detail = 51;
+
+  // For PyTorch serialized tensor.
+  optional int64 require_gradient = 52;
+  optional int64 is_buffer = 53;
+}
+
+// Defines a tensor shape. A dimension can be either an integer value
+// or a symbolic variable. A symbolic variable represents an unknown
+// dimension.
+message TensorShapeProto {
+  message Dimension {
+    oneof value {
+      int64 dim_value = 1;
+      string dim_param = 2;   // namespace Shape
+    };
+    // Standard denotation can optionally be used to denote tensor
+    // dimensions with standard semantic descriptions to ensure
+    // that operations are applied to the correct axis of a tensor.
+    // Refer to https://github.com/onnx/onnx/blob/master/docs/DimensionDenotation.md#denotation-definition
+    // for pre-defined dimension denotations.
+    optional string denotation = 3;
+  };
+  // To represent a scalar, using no dim to represent 0-d tensor.
+  repeated Dimension dim = 1;
+
+  repeated Dimension stride = 51;
+}
+
+// Types
+//
+// The standard ONNX data types.
+message TypeProto {
+
+  message Tensor {
+    // This field MUST NOT have the value of UNDEFINED
+    // This field MUST be present for this version of the IR.
+    optional TensorProto.DataType elem_type = 1;
+    optional TensorShapeProto shape = 2;
+  }
+
+  // Sequence type: List, Tuple
+  message Sequence {
+    // elem_type and elem_type_list cannot appear together.
+    // If all the element types are the same, we use elem_type,
+    // otherwise, we specify the type of each element in elem_type_list.
+    optional TypeProto elem_type = 1;
+    repeated TypeProto elem_type_list = 51;
+    enum SequenceType {
+      UNDEFINED = 0;
+      LIST = 1;
+      TUPLE = 2;
+    }
+    optional SequenceType sequence_type = 52;
+  }
+
+  // Map<K, V>, (not necessary at this moment)
+  message Map {
+    optional TensorProto.DataType key_type = 1;
+    optional TypeProto value_type = 2;
+  }
+
+  // Special type of blobs, based on the type_name, we can choose the right
+  // serializer and deserialzier.
+  message SpecialBlob {
+    optional string type_name = 1;
+  }
+
+  oneof value {
+    // The type of a tensor.
+    Tensor tensor_type = 1;
+    Sequence sequence_type = 4;
+    Map map_type = 5;
+    SpecialBlob special_type = 51;
+  }
+
+  // An optional denotation can be used to denote the whole 
+  // type with a standard semantic description as to what is 
+  // stored inside. Refer to https://github.com/onnx/onnx/blob/master/docs/TypeDenotation.md#type-denotation-definition
+  // for pre-defined type denotations.
+  optional string denotation = 6;
+}
+
+// TensorProtos stores multiple TensorProto objects in one single proto. This
+// is useful for small tensors; For anything big, consider using a DB for
+// storage.
+message TensorProtos {
+  repeated TensorProto protos = 1;
```

Comment:
Just wondering, what's the usecase for this?

Reply:
It's for Caffe2, they are used at several places, for example:
https://github.com/pytorch/pytorch/blob/cca247635c6edb323176eeac7a18d3e9ab71c558/caffe2/operators/tensor_protos_db_input.cc